[WIP] changefeedccl/kvevent: replace the memBuffer with a blockingBuffer

ajwerner opened the pull request:
changefeedccl/kvevent: replace the memBuffer with a blockingBuffer
    
    The memBuffer, for historical reasons, would use row buffering technology to
    store kvfeed events. The idea behind that was that maybe one day we might want
    to spill to disk (I don't think since the advent of the rangefeed that we ever
    would) and that that code already dealt with the memory monitoring. All-in-all
    the code was lazy and had the fatal flow that bursts of events from the
    rangefeeds could lead to changefeeds which would normally surive to fail.
    
    This PR replaces that implementation with a simpler, self-contained
    implementation which will block when there is insufficient available
    memory as opposed to failing. It will fail if any individual row exceeds
    the total capacity which can be allocated.
    
    Release note (enterprise change): Improved the internal bufferring of
    CHANGEFEEDs to be robust to bursts in traffic which exceed the throughput
    to the sink.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63050)
<!-- Reviewable:end -->

ajwerner committed:
changefeedccl/kvevent: extract subpackage from kvfeed

The Event code was pretty self-contained. It feels ripe for extraction in order
to improve testing and clarify concept. The kvfeed package depends on this new
package. This commit is purely code movement.

Release note: None

ajwerner committed:
changefeedccl: remove a literally unused metric

Release note: None

ajwerner committed:
changefeedccl/kvevent: refactor event constructors

Release note: None

ajwerner committed:
changefeedccl/kvevent: replace the memBuffer with a blockingBuffer

The memBuffer, for historical reasons, would use row buffering technology to
store kvfeed events. The idea behind that was that maybe one day we might want
to spill to disk (I don't think since the advent of the rangefeed that we ever
would) and that that code already dealt with the memory monitoring. All-in-all
the code was lazy and had the fatal flow that bursts of events from the
rangefeeds could lead to changefeeds which would normally surive to fail.

This PR replaces that implementation with a simpler, self-contained
implementation which will block when there is insufficient available
memory as opposed to failing. It will fail if any individual row exceeds
the total capacity which can be allocated.

Release note (enterprise change): Improved the internal bufferring of
CHANGEFEEDs to be robust to bursts in traffic which exceed the throughput
to the sink.

ajwerner commented:
I think this is what we want.

ajwerner commented:
@miretskiy is this something we'd want to merge if I spruced it up or should I just close it?

ajwerner commented:
Superseded by https://github.com/cockroachdb/cockroach/pull/66503.
